CMakeLists.txt

```diff
@@ -168,15 +185,24 @@ endif()
 if(USE_CUDA)
   set(USE_OPENMP ON CACHE BOOL "CUDA requires OpenMP" FORCE)
   # `export CXX=' is ignored by CMake CUDA.
-  set(CMAKE_CUDA_HOST_COMPILER ${CMAKE_CXX_COMPILER})
+  set(CMAKE_CUDA_HOST_COMPILER ${CMAKE_CXX_COMPILER} CACHE FILEPATH
+    "The compiler executable to use when compiling host code for CUDA or HIP language files.")
+  mark_as_advanced(CMAKE_CUDA_HOST_COMPILER)
   message(STATUS "Configured CUDA host compiler: ${CMAKE_CUDA_HOST_COMPILER}")
 
+  set(CMAKE_CUDA_RUNTIME_LIBRARY Static CACHE STRING
```

Comment:
Don't we just want to specify a value for this after `enable_language(CUDA)` or is this fixing some bug I am not aware of?

Reply:
It's currently explicitly being set as `Static` directly in the target property so making it a cache variable defaulting to `Static` preserves the existing behavior but lets the user pass in `-DCMAKE_CUDA_RUNTIME_LIBRARY=Shared` if they want to (we do :-) ). If it waits until after `enable_language` you wouldn't be able to tell if the value being used is user specified (which we want to keep) or a detected default (which we want to override with `Static` to maintain current behavior). At least I think so?

Reply:
You should be able to remove the cache variable entirely and just use `-DCMAKE_CUDA_RUNTIME_LIBRARY=Shared` . The internal default that cmake computes will go into a different variable and should be used only if no user value for `CMAKE_CUDA_RUNTIME_LIBRARY` exists.

Reply:
I now see where xgboost was explicitly using the per property var., I would go with:
```
if(NOT DEFINED CMAKE_CUDA_RUNTIME_LIBRARY)
  set(CMAKE_CUDA_RUNTIME_LIBRARY Static)
endif()
```
That way we don't influence other projects that consume xgboost via a source build ( FetchContent, CPM, etc ) with a cache variable.

The `CMAKE_CUDA_HOST_COMPILER` is a different since it is used during compiler detection, and therefore needs to be consistent across all targets. I am okay with the current approach for that.